File: Meadow_DotNet_SDK/Project_Templates/templates/MeadowFeatherVBNet/MeadowApp.vb
﻿Imports System
Imports System.Threading.Tasks
Imports Meadow
Imports Meadow.Devices
Imports Meadow.Foundation.Leds
Imports Meadow.Peripherals.Leds

Public Class MeadowApp
    Inherits App(Of F7FeatherV2)

    Private onboardLed As RgbPwmLed

    Public Overrides Function Initialize() As Task
        Resolver.Log.Info("Initialize...")

        onboardLed = New RgbPwmLed(
            Device.Pins.OnboardLedRed,
            Device.Pins.OnboardLedGreen,
            Device.Pins.OnboardLedBlue,
            CommonType.CommonAnode)

        Return Task.CompletedTask
    End Function

    Public Overrides Async Function Run() As Task
        Resolver.Log.Info("Run...")

        Await CycleColors(TimeSpan.FromMilliseconds(1000))

    End Function

    Private Async Function CycleColors(ByVal duration As TimeSpan) As Task
        Resolver.Log.Info("Cycle colors...")

        While True
            Await ShowColorPulse(Color.Blue, duration)
            Await ShowColorPulse(Color.Cyan, duration)
            Await ShowColorPulse(Color.Green, duration)
            Await ShowColorPulse(Color.GreenYellow, duration)
            Await ShowColorPulse(Color.Yellow, duration)
            Await ShowColorPulse(Color.Orange, duration)
            Await ShowColorPulse(Color.OrangeRed, duration)
            Await ShowColorPulse(Color.Red, duration)
            Await ShowColorPulse(Color.MediumVioletRed, duration)
            Await ShowColorPulse(Color.Purple, duration)
            Await ShowColorPulse(Color.Magenta, duration)
            Await ShowColorPulse(Color.Pink, duration)
        End While

    End Function

    Private Async Function ShowColorPulse(ByVal color As Color, ByVal duration As TimeSpan) As Task
        Await onboardLed.StartPulse(color, duration / 2)
        Await Task.Delay(duration)
    End Function

End Class

File: Meadow_DotNet_SDK/Project_Templates/templates/MeadowFeatherFSharp/MeadowApp.fs
﻿namespace MeadowApplication.Template

open System
open Meadow
open Meadow.Devices
open Meadow.Foundation.Leds
open Meadow.Peripherals.Leds
open System.Threading.Tasks

type MeadowApp() =
    inherit App<F7FeatherV2>()

    let mutable led : RgbPwmLed = null

    let ShowColorPulse (color : Color) (duration : TimeSpan) = async {
        do! led.StartPulse(color, TimeSpan.FromMilliseconds(500.0)) |> Async.AwaitTask
        do! Async.Sleep duration
        do! led.StopAnimation() |> Async.AwaitTask
    }

    let CycleColors (duration : TimeSpan) = async {
        do Resolver.Log.Info "Cycle colors..."

        while true do
            do! ShowColorPulse Color.Blue duration 
            do! ShowColorPulse Color.Cyan duration
            do! ShowColorPulse Color.Green duration
            do! ShowColorPulse Color.GreenYellow duration
            do! ShowColorPulse Color.Yellow duration
            do! ShowColorPulse Color.Orange duration
            do! ShowColorPulse Color.OrangeRed duration
            do! ShowColorPulse Color.Red duration
            do! ShowColorPulse Color.MediumVioletRed duration
            do! ShowColorPulse Color.Purple duration
            do! ShowColorPulse Color.Magenta duration
            do! ShowColorPulse Color.Pink duration
    }

    override this.Initialize() =
        do Resolver.Log.Info "Initialize..."

        led <- new RgbPwmLed(
            MeadowApp.Device.Pins.OnboardLedRed,
            MeadowApp.Device.Pins.OnboardLedGreen, 
            MeadowApp.Device.Pins.OnboardLedBlue, 
            CommonType.CommonAnode)

        Task.CompletedTask;

    override this.Run () : Task =
        let runAsync = async {
            do Resolver.Log.Info "Run..."
            do! CycleColors(TimeSpan.FromSeconds(1.0))
        }
        Async.StartAsTask(runAsync) :> Task